net/instaweb/rewriter/webscale_make_scripts_async.cc

```diff
@@ -0,0 +1,108 @@
+#include "net/instaweb/rewriter/public/rewrite_driver_factory.h"
+#include "net/instaweb/rewriter/public/webscale_make_scripts_async.h"
+#include "pagespeed/kernel/html/html_element.h"
+#include "pagespeed/kernel/html/html_name.h"
+#include "pagespeed/kernel/html/html_node.h"
+#include "pagespeed/kernel/base/message_handler.h"
+#include "pagespeed/kernel/base/statistics.h"
+#include "pagespeed/kernel/base/string.h"
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/kernel/util/re2.h"
+
+namespace net_instaweb {
+
+WebscaleMakeScriptsAsync::WebscaleMakeScriptsAsync(RewriteDriver* rewrite_driver, MessageHandler* message_handler)
+    : CommonFilter(rewrite_driver) {
+  Statistics* statistics = rewrite_driver->statistics();
+  message_handler = driver()->message_handler();
+  escaped_urls = ConstructPatternFromCustomUrls(rewrite_driver->options());
+}
+
+WebscaleMakeScriptsAsync::~WebscaleMakeScriptsAsync(){
+}
+
+void WebscaleMakeScriptsAsync::InitStats(Statistics* statistics) {
+}
+
+
+void WebscaleMakeScriptsAsync::StartDocumentImpl() {
+}
+
+
+void WebscaleMakeScriptsAsync::StartElementImpl(HtmlElement* element) {
+  message_handler = driver()->message_handler();
+
+  // Script element found.
+  if (element->keyword() == HtmlName::kScript) {
+    const char* src_attribute = element->EscapedAttributeValue(HtmlName::kSrc);
+    // Script element has a 'src' attribute.
+    if (src_attribute != NULL) {
+      // Convert the src element to a GoogleString format for compatibility.
+      GoogleString* src_string = new GoogleString(src_attribute);
+
+      // If there are no custom urls mentioned, print a message and do nothing.
+      if (escaped_urls == "") {
+        message_handler->Message(
+          kInfo, "No custom urls provided.");
+      }
+      else {
+        // A full match of the custom url needs to be found.
+        bool match = RE2::FullMatch(src_string->c_str(), escaped_urls.c_str());
+        if (match) {
+          // If a match is found, add the async attribute.
+          driver()->AddAttribute(element, HtmlName::kAsync, "true");
+          // Set the debug comment so that it will be displayed when
+          // ModPagespeedEnabledFilters=+debug is used.
+          driver()->InsertDebugComment("Webscale added an async attribute successfully", element);
+          message_handler->Message(
+            kInfo, "Adding an async attribute to %s.", src_string->c_str());
+        } else {
+            message_handler->Message(
+              kInfo, "Not adding an async attribute to %s.", src_string->c_str());
+        }
+      }
+    }
+  }
+}
+
+// Construct a regular expressions with the custom urls provied. Each custom
+// url will be escaped and an OR of all the escaped urls will be constructed
+// for pattern matching.  Pattern matching is preferred here instead of
+// iterating the list of cusotm urls every time a src attribute is encoutered.
+// This make comparison faster.
```

Comment:
except that regexp matching is much slower than a std::set lookup.... right?  At least, I think -- I haven't benchmarked, but I'd be pretty surprised if regex was faster, and std::set should be simpler as well.  This is especially true since you are parsing and compiling the regular expression each time.

```c++
std::set<GoogleString> WebscaleMakeScriptsAsync::MakeSet(const RewriteOptions* options) {
  std::set<GoogleString> urls;
  for (int i = 0; i < options->num_custom_async_urls(); i++) {
    urls.insert(options->custom_async_url(i));
  }
  return urls;
}
```

And then when it's used:

```c++
if (custom_urls.find(src_attribute) != custom_urls.end()) {
  // found!
}
```